File: python/day02.py
from pathlib import Path
from dataclasses import dataclass, field

data = (Path(__file__).parent.parent / "data" / "day02.txt").read_text()

RED = 12
GREEN = 13
BLUE = 14

@dataclass
class Round:
    blue: int = 0
    red: int = 0
    green: int = 0


@dataclass
class Game:
    rounds: list[Round] = field(default_factory=list)


def parse_lines(data: str):
    games = []
    for row in data.splitlines():
        game = Game()
        a, b = row.split(":")
        a = int(a.lstrip("Game "))
        b = b.split(';')
        for round in b:
            roundmodel = Round()
            plays = round.strip().split(",")
            for colors in plays:
                count, color = colors.split()
                match color:
                    case 'blue':
                        roundmodel.blue = int(count)
                    case 'red':
                        roundmodel.red = int(count)
                    case 'green':
                        roundmodel.green = int(count)
                game.rounds.append(roundmodel)
        games.append(game)
    return games


def is_possible(game: Game):
    for round in game.rounds:
        if round.blue > BLUE:
            return False
        if round.green > GREEN:
            return False
        if round.red > RED:
            return False
    return True


def count_possible(games: list[Game]):
    return sum(i for i, game in enumerate(games, 1) if is_possible(game))


def power(game: Game):
    red = max(x.red for x in game.rounds)
    green = max(x.green for x in game.rounds)
    blue = max(x.blue for x in game.rounds)
    return red * green * blue


if __name__ == "__main__":
    games = parse_lines(data)
    count = count_possible(games)
    print(count)
    powers = sum(power(game) for game in games)
    print(powers)

File: rust/src/bin/day02.rs
use std::fs;

type Num = i32;

const RED: Num = 12;
const GREEN: Num = 13;
const BLUE: Num = 14;

#[derive(Debug)]
struct Round {
    red: Num,
    green: Num,
    blue: Num,
}

#[derive(Debug)]
struct Game {
    rounds: Vec<Round>,
}

fn main() {
    let raw = fs::read_to_string("../data/day02.txt").unwrap();
    let games = parse(&raw);

    let p1 = count_possible(&games);
    dbg!(p1);
    let p2: Num = games.iter().map(|game| power(game)).sum();
    dbg!(p2);
}

fn is_possible(game: &Game) -> bool {
    for round in game.rounds.iter() {
        if round.blue > BLUE {
            return false;
        }
        if round.green > GREEN {
            return false;
        }
        if round.red > RED {
            return false;
        }
    }
    return true;
}

fn power(game: &Game) -> Num {
    let red: Num = game.rounds.iter().map(|x| x.red).max().unwrap();
    let green: Num = game.rounds.iter().map(|x| x.green).max().unwrap();
    let blue: Num = game.rounds.iter().map(|x| x.blue).max().unwrap();
    return red * green * blue;
}

fn count_possible(games: &Vec<Game>) -> Num {
    return games
        .iter()
        .enumerate()
        .filter_map(|(i, game)| {
            if is_possible(game) {
                return Some(i as Num);
            }
            return None;
        })
        .sum();
}

fn parse(data: &str) -> Vec<Game> {
    return data
        .trim()
        .lines()
        .map(|line| {
            let mut split = line.split(":");
            split.next();
            let b = split.next().unwrap();
            let rounds = b
                .split(";")
                .map(|part| {
                    let mut round = Round {
                        red: 0,
                        green: 0,
                        blue: 0,
                    };
                    for colors in part.trim().split(",") {
                        let mut colorsplit = colors.trim().split(" ");
                        let count: Num = colorsplit.next().unwrap().parse().unwrap();
                        let color = colorsplit.next().unwrap();
                        match color {
                            "blue" => round.blue = count,
                            "red" => round.red = count,
                            "green" => round.green = count,
                            _ => {
                                unreachable!("Uh oh");
                            }
                        }
                    }
                    return round;
                })
                .collect::<Vec<Round>>();
            return Game { rounds };
        })
        .collect();
}
